init.lua

```diff
@@ -146,19 +175,30 @@ function cudnn.toDescriptor(t)
    return descriptor
 end
 
+function cudnn.createDescriptors(count, descs_type, create_func, destroy_func)
+   local ds = ffi.new(descs_type, count)
+   for i = 0, count - 1 do
+      errcheck(create_func, ds + i)
+   end
+   local function destroyDescriptors(ds)
+      for i = 0, count - 1 do
+         errcheck(destroy_func, ds[i])
+      end
+   end
+   ffi.gc(ds, destroyDescriptors)
+   return ds
+end
 
 local sharedBuffer = {}
-for i=1,numDevices do
-    sharedBuffer[i] = {}
-end
 
 function cudnn.getSharedWorkspace()
     local device = cutorch.getDevice()
-    local stream = cutorch.getStream() -- starts from 0
-    if not sharedBuffer[device][stream] then
-       sharedBuffer[device][stream] = torch.CudaTensor(1)
+    if not sharedBuffer[device] then
```

Comment:
I dont think we should add a memory pool. Especially not like THCGetTempBuf.
@colesbury is working this week on a custom memory allocator that will be integrated into cutorch, so that we can do malloc and free without dealing with sync points.
We looked at CUB and CNMem, but both fall short on a few aspects that we want, so he's working on something slightly different, but starting from CUB / CNMem.
If there's anyone at NVIDIA we should talk to as he works on this, we'd love to get thoughts.
We want to build a drop-in replacement which doesn't have the requirement of specifying before-hand how much memory is being managed, so that this change is completely transparent to the user.